### PR TITLE
Objects exercise - 06.storeCatalogue solution

### PR DESCRIPTION
**Pull request by NStoyanov95**
exercise solution

### CODE REVIEW

**Review by Daooobg — APPROVED** [review]
Very short and clean code

**Comment by zlatozaraZlatkova** [comment]
Great job! My approach would be to tweak the code to make it impossible to create an item that is already on the list, as is currently permitted, and to update the latest incoming prices as iterates.

Suggestion for optimization: 
1. create a new Map object that only collects unique products with their last updated price. 
2. iterate through the sorted key-value pairs and print the result.

**Review by Daooobg — APPROVED** [review]

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+function solve(arr) {
+
```
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:2]:
```suggestion
const catalogue = new Map();
```

Comment (zlatozaraZlatkova) [reply 03-objects-and-composition/exercise/06.storeCatalogue.js:2]: Create a new Map() object that only collects unique products with their last updated price.

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    arr.sort((a, b) => a.localeCompare(b));
```
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:3]:
```suggestion
    //arr.sort((a, b) => a.localeCompare(b));
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    let currentLetter;
+
+    arr.forEach(el => {
+        if (el[0][0] != currentLetter) {
+            currentLetter = el[0][0];
+            console.log(currentLetter);
+        }
+
+        let [product, price] = el.split(' : ');
+        console.log(`  ${product}: ${price}`);
```
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:14]:
```suggestion
 let [product, price] = el.split(' : ');
      if (!catalogue.has(product)) {
        catalogue.set(product, price);
      } 
      catalogue.get(product, price);
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    });
```
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:15]:
```suggestion
    });
     let sortProducts = Array.from(catalogue.entries()).sort((a, b) => a[0].localeCompare(b[0]));

  for (let [product, price] of sortProducts) {
    getFirstLetter(product);
    if(currentLetter === product[0]) {
      console.log(`  ${product}: ${price}`);
    }
    
  }

  function getFirstLetter(item) {
    if (item[0] !== currentLetter) {
      currentLetter = item[0];
      console.log(currentLetter);
      return currentLetter
    }
  }
```

**Review by zlatozaraZlatkova — COMMENTED** [review]
Great job!

My approach would be to tweak the code to make it impossible to create an item that is already on the list, as is currently permitted, and to update the latest incoming prices as iterates.
Suggestion for optimization:

- create a new Map object that only collects unique products with their last updated price.
- iterate through the sorted key-value pairs and print the result.

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 4 above)
Comment (NStoyanov95) [reply 03-objects-and-composition/exercise/06.storeCatalogue.js:15]: This is the right way to solve the problem related to the topic we're discussing in class.

**Comment by NStoyanov95** [comment]
> Great job!
> 
> My approach would be to tweak the code to make it impossible to create an item that is already on the list, as is currently permitted, and to update the latest incoming prices as iterates. Suggestion for optimization:
> 
> * create a new Map object that only collects unique products with their last updated price.
> * iterate through the sorted key-value pairs and print the result.

Great, this is the right way to solve the problem related to the topic we're discussing in class

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 2 above)
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:3]:
```suggestion
   //arr.sort((a, b) => a.localeCompare(b));
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 1 above)
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:2]:
```suggestion
 const catalogue = new Map();
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 3 above)
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:14]:
```suggestion
 let [product, price] = el.split(' : ');
      if (!catalogue.has(product)) {
        catalogue.set(product, price);
      } 
      catalogue.get(product, price);
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 4 above)
Comment (zlatozaraZlatkova) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:15]:
```suggestion
});

let sortProducts = Array.from(catalogue.entries()).sort((a, b) =>
    a[0].localeCompare(b[0])
  );

  for (let [product, price] of sortProducts) {
    getFirstLetter(product);
    if (currentLetter === product[0]) {
      console.log(`  ${product}: ${price}`);
    }
  }

  function getFirstLetter(item) {
    if (item[0] !== currentLetter) {
      currentLetter = item[0];
      console.log(currentLetter);
      return currentLetter;
    }
  }
```

File: 03-objects-and-composition/exercise/06.storeCatalogue.js (hunk 2 above)
Comment (antoniaat) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:3]: I would advice you to add a variable for the sorted data and not modify the passed parameter:
 `const sortedInput = arr.sort((a, b) => a.localeCompare(b));`

You can read this topic about it: https://medium.com/the-andela-way/its-bad-practice-to-reassign-function-params-c37f289a5dc4

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+function solve(arr) {
+
+    arr.sort((a, b) => a.localeCompare(b));
+
+    let currentLetter;
```
Comment (antoniaat) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:5]: Why defining the variable here and not in the forEach?

File: 03-objects-and-composition/exercise/06.storeCatalogue.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    arr.forEach(el => {
+        if (el[0][0] != currentLetter) {
```
Comment (antoniaat) [inline 03-objects-and-composition/exercise/06.storeCatalogue.js:8]: I would suggest you to add more descriptive names to the variables and extract more items like el[0][0].

Comment (NStoyanov95) [reply 03-objects-and-composition/exercise/06.storeCatalogue.js:8]: @antoniaat,  Thanks for the advice.

**Review by MilenaGeorgieva95 — APPROVED** [review]

**Pull request merged**